glusterd2/commands/snapshot/snapshot-create.go

```diff
@@ -390,50 +390,73 @@ func validateSnapCreate(c transaction.TxnCtx) error {
 	return nil
 }
 func takeVolumeSnapshots(newVol, oldVol *volume.Volinfo) error {
+	var wg sync.WaitGroup
+	numBricks := len(oldVol.GetBricks())
+	errCh := make(chan error, numBricks)
 	for subvolCount, subvol := range oldVol.Subvols {
 		for count, b := range subvol.Bricks {
 			if !uuid.Equal(b.PeerID, gdctx.MyUUID) {
 				continue
 			}
-			/*
-				TODO : Run as a go routine
-			*/
+			wg.Add(1)
 			snapBrick := newVol.Subvols[subvolCount].Bricks[count]
-			mountData := snapBrick.MountInfo
-			length := len(b.Path) - len(mountData.Mountdir)
-			mountRoot := b.Path[:length]
-			mntInfo, err := volume.GetBrickMountInfo(mountRoot)
-			if err != nil {
-				return err
-			}
+			go brickSnapshot(errCh, &wg, snapBrick, b)
 
-			log.WithFields(log.Fields{
-				"mount device": mntInfo.FsName,
-				"devicePath":   mountData.DevicePath,
-				"Path":         b.Path,
-			}).Debug("Running snapshot create command")
+		}
+	}
+	go func() {
+		wg.Wait()
```

Comment:
Waiting on goroutines by launching another goroutine does not seem like a good idea. You should wait in a goroutine that you have control over.

I suggest that you collect the errors in a different goroutine (the loop below) and wait for the snapshot goroutines and the err collect goroutine in this function.

Reply:
done